Framework/Beamline/src/ComponentInfo.cpp

```diff
@@ -14,38 +15,179 @@ ComponentInfo::ComponentInfo()
+      m_assemblySortedComponentIndices->begin() + range.second);
+}
+
 size_t ComponentInfo::size() const { return m_size; }
 
 bool ComponentInfo::operator==(const ComponentInfo &other) const {
-  return m_size == other.m_size && m_ranges == other.m_ranges &&
+  return m_size == other.m_size && m_detectorRanges == other.m_detectorRanges &&
          *m_assemblySortedDetectorIndices ==
              *other.m_assemblySortedDetectorIndices;
 }
 bool ComponentInfo::operator!=(const ComponentInfo &other) const {
   return !(this->operator==(other));
 }
+
+Eigen::Vector3d ComponentInfo::position(const size_t componentIndex) const {
+  if (isDetector(componentIndex)) {
+    return m_detectorInfo->position(componentIndex);
+  }
+  const auto rangesIndex = compOffsetIndex(componentIndex);
+  return (*m_positions)[rangesIndex];
+}
+
+Eigen::Quaterniond ComponentInfo::rotation(const size_t componentIndex) const {
+  if (isDetector(componentIndex)) {
+    return m_detectorInfo->rotation(componentIndex);
+  }
+  const auto rangesIndex = compOffsetIndex(componentIndex);
+  return (*m_rotations)[rangesIndex];
+}
+
+/**
+ * Sets the rotation for a component described by target component index
+ *
+ * This will propagate and apply the derived position offsets to all known
+ *sub-components
+ *
+ * @param componentIndex : Component index to update at
+ * @param newPosition : Absolute position to set
+ */
+void ComponentInfo::setPosition(const size_t componentIndex,
+                                const Eigen::Vector3d &newPosition) {
+
+  const Eigen::Vector3d offset = newPosition - position(componentIndex);
+  const auto indices = this->componentIndices(
+      componentIndex); // Includes requested component index
+  for (auto &index : indices) {
+    if (isDetector(index)) {
+      m_detectorInfo->setPosition(index,
+                                  m_detectorInfo->position(index) + offset);
+    } else {
+      m_positions.access()[compOffsetIndex(index)] += offset;
+      // TODO. This could change L1/L2. Is that being detected?
+    }
+  }
+}
+
+/**
+ * Sets the rotation for a component described by target component index.
+ *
+ * This will propagate and apply the derived rotation to all known
+ *sub-components
+ * This will also update derived positions for target component and all
+ *sub-components
+ *
+ * @param componentIndex : Component index to update at
+ * @param newRotation : Absolute rotation to set
+ */
+void ComponentInfo::setRotation(const size_t componentIndex,
+                                const Eigen::Quaterniond &newRotation) {
+
+  using namespace Eigen;
+  const Eigen::Vector3d compPos = position(componentIndex);
+  const Eigen::Quaterniond currentRotInv = rotation(componentIndex).inverse();
+  const Eigen::Quaterniond rotDelta =
+      (newRotation * currentRotInv).normalized();
+
+  const Affine3d transform =
+      Translation3d(compPos) * rotDelta * Translation3d(-compPos);
```

Comment:
I think I had tried using `Affine3d` but in the end it was faster to use a manual position shift with a `Matrix3d` (see `API::DetectorInfo`). This is performance critical for `SCDCalibratePanels`, but maybe you can double check.